Source/WebKit/Platform/IPC/Connection.h

```diff
@@ -650,16 +675,36 @@ Connection::AsyncReplyID Connection::sendWithAsyncReply(T&& message, C&& complet
     return { };
 }
 
+template<typename T, typename C>
+Connection::AsyncReplyID Connection::sendWithAsyncReplyOnDispatcher(T&& message, RefCountedSerialFunctionDispatcher& dispatcher, C&& completionHandler, uint64_t destinationID, OptionSet<SendOption> sendOptions)
+{
+    static_assert(!T::isSync, "Async message expected");
+    auto handler = makeAsyncReplyHandlerWithDispatcher<T>(std::forward<C>(completionHandler), dispatcher);
+    auto replyID = handler.replyID;
+    auto encoder = makeUniqueRef<Encoder>(T::name(), destinationID);
+    encoder.get() << message.arguments();
+    if (sendMessageWithAsyncReplyWithDispatcher(WTFMove(encoder), WTFMove(handler), sendOptions) == Error::NoError)
+        return replyID;
+    // FIXME: Propagate the error back.
+    return { };
+}
+
 template<typename T>
 Ref<typename T::Promise> Connection::sendWithPromisedReply(T&& message, uint64_t destinationID, OptionSet<SendOption> sendOptions)
+{
+    return sendWithPromisedReplyOnDispatcher<T>(std::forward<T>(message), RunLoop::main(), destinationID, sendOptions);
```

Comment:
In the previous implementation, the promise was resolved on the  dispatcher of the connection (that is passed to connection via open()). This is a change of functionality to use the main runloop.

Reply:
Good point,
All users of sendWithPromisedReply where using the GPUConnection's `Connection` which is the main thread, so the change is equivalent functionally.

In addition, where the promise is resolved is irrelevant really. Ideally it would be a dedicated RefCountedSerialFunctionDispatcher, IPC's WorkQueue would have been ideal, but considering we don't want to run the IPC Decoder to run on the IPC WorkQueue, it provides some challenges in deciding what is the best course of action.

Maybe I should simply create a 2nd Workqueue just for resolving the promise. WorkQueue on Cocoa is very light weight, and will likely not even create a new thread.

Reply:
I reverted that change and now continue to use the Connection's dispatcher to resolve the promise when using `sendWithPromisedReply`.

Good catch, thanks, it caused failures in the API tests.